Neos.Neos/Classes/Domain/Service/WorkspaceService.php

```diff
@@ -0,0 +1,316 @@
+
+        return WorkspacePermissions::create(
+            read: $userIsAdmin || $userWorkspaceRole->isAtLeast(WorkspaceRole::COLLABORATOR),
+            write: $userIsAdmin || $userWorkspaceRole->isAtLeast(WorkspaceRole::COLLABORATOR),
+            manage: $userIsAdmin || $userWorkspaceRole->isAtLeast(WorkspaceRole::MANAGER),
+        );
+    }
+
+    private function findPrimaryWorkspaceNameForUser(ContentRepositoryId $contentRepositoryId, UserId $userId): ?WorkspaceName
+    {
+        $table = self::TABLE_NAME_WORKSPACE_ROLE;
+        $query = <<<SQL
+            SELECT
+                workspace_name
+            FROM
+                {$table}
+            WHERE
+                content_repository_id = :contentRepositoryId
+                AND subject = :subject
+                AND role = :ownerRole
+        SQL;
+        $workspaceName = $this->dbal->fetchOne($query, [
+            'contentRepositoryId' => $contentRepositoryId->value,
+            'subject' => 'user:' . $userId->value,
+            'ownerRole' => 'OWNER'
+        ]);
+        return $workspaceName === false ? null : WorkspaceName::fromString($workspaceName);
+    }
+
+    /**
+     * @param array<string> $userRoles
+     */
+    private function getWorkspaceRoleOfUser(ContentRepositoryId $contentRepositoryId, WorkspaceName $workspaceName, UserId $userId, array $userRoles): WorkspaceRole
+    {
+        $table = self::TABLE_NAME_WORKSPACE_ROLE;
+        $query = <<<SQL
+            SELECT
+                role
+            FROM
+                {$table}
+            WHERE
+                content_repository_id = :contentRepositoryId
+                AND workspace_name = :workspaceName
+                AND subject IN (:subjects)
+            ORDER BY
+                CASE
+                    WHEN role='OWNER' THEN 1
+                    WHEN role='MANAGER' THEN 2
+                    WHEN role='COLLABORATOR' THEN 3
+                END
+            LIMIT 1
+        SQL;
+        $role = $this->dbal->fetchOne($query, [
+            'contentRepositoryId' => $contentRepositoryId->value,
+            'workspaceName' => $workspaceName->value,
+            'subjects' => ['user:' . $userId->value, ...array_map(static fn (string $role) => 'group:' . $role, $userRoles)],
+        ], [
+            'subjects' => Connection::PARAM_STR_ARRAY,
+        ]);
+        if ($role === false) {
+            return WorkspaceRole::NONE;
+        }
+        return WorkspaceRole::from($role);
+    }
+
+    private static function getUniqueWorkspaceName(ContentRepository $contentRepository, string $candidate): WorkspaceName
+    {
+        $workspaceNameCandidate = WorkspaceName::transliterateFromString($candidate);
+        $workspaceName = $workspaceNameCandidate;
+        while ($contentRepository->getWorkspaceFinder()->findOneByName($workspaceName) instanceof Workspace) {
+            $workspaceName = WorkspaceName::fromString(
```

Comment:
with the lowercase enforcement `generateRandomString` is not safe and will crash